erpnext/stock/doctype/bin/bin.py

```diff
@@ -94,6 +94,28 @@ def update_reserved_qty_for_production(self):
 			self.db_set('reserved_qty_for_production', self.reserved_qty_for_production)
 			self.db_set('projected_qty', self.projected_qty)
 
+	def update_reserved_qty_for_sub_contracting(self, po_name, transferred_qty, transaction_type):
+		#Update Reserved Quantity for Sub Contracting in Bin
+		if transaction_type == "Reserve":
+			required_qty = frappe.db.sql('''select sum(itemsup.required_qty)
+				from `tabItem` item, `tabPurchase Order` po, `tabPurchase Order Item Supplied` itemsup
+				where
+					item.name = itemsup.rm_item_code
+					and po.name = %s
+					and itemsup.rm_item_code = %s
+					and itemsup.parent = po.name
+					and po.docstatus = 1
+					and po.is_subcontracted = 'Yes'
+					and item.default_warehouse = %s''', (po_name, self.item_code, self.warehouse))[0][0]
```

Comment:
We cannot rely on default warehouse set in Item master. It can be changed after submitting PO, and if the user cancelled thereafter what will happen? We need to introduce a new field for "Reserve Warehouse" inside "PO Item"/ "Item Supplied" table, we can set the default value based on item.

Also this type of function should not be dependent on args like po_name, transferred_qty, transaction_type. The function should be self dependent, so that we can call the function from other places like patches etc.
So, basically, it should recalculate reserved qty for the item and warehouse based on all submitted Purchase Orders and Subcontracted Stock Entries.

Reply:
sorry @nabinhait  but when I make a Production Order the material is reserved in the "Source Warehouse", you mean a new field "Source Warehouse" shall be created also in Subcontracted Purchase Order? 
Because in Production Order we actually don't have the "Reserve Warehouse" field.
Anyway if I actually make a Production Order and reserve material in a specific warehouse I still can transfer material from another warehouse and reserved material is deduced the same.